[DM-31792] Initial draft

rra opened the pull request.

rra committed:
Initial version

frossie committed:
minor edits

rra committed:
Merge pull request #1 from lsst-dm/u/fe/1

minor edits

rra committed:
Clarify Butler URLs further

Explicitly mention signed URLs and where the result URLs are
stored.

rra committed:
Typo fix

rra committed:
Add XML handling section

rra committed:
Flesh out with more details from my notes

Discuss aborting jobs, quotas and throttling, and some additional
details of the API service.  Add an open questions section.

rra committed:
Add discussion of alternate image formats

rra committed:
Document use of the Butler for object storage

One implication is the need for Content-Type metadata.  Document
that.

rra committed:
Restrict the type of filtering parameters

rra committed:
Updates based on Thursday's discussion

Major changes include:

- Reorganize and clarify the results and alternate image format
  handling.
- Be explicit about wanting long-running backend workers.
- Allow multiple ID filter parameters.
- Discuss discovery issues and the expected ObsTAP implementation.

rra committed:
Fix broken numbered list

rra committed:
Fix one confusion between task and pipeline task

rra committed:
Alternate image type wording fixes

rra committed:
Fix footnote syntax

rra committed:
Fix minor POS syntax issue

rra committed:
More wording cleanup

There were some remaining problems in the results section.

rra committed:
Add another paragraph break

rra committed:
Mention storing results in VOSpace

rra committed:
The Butler collection must be a URL

The IVOA UWS standard requires that the result be a URL.

timj commented on index.rst:
```diff
@@ -1,61 +1,333 @@
+The result of a sync request that does not request an alternate image format is the FITS file.
+Therefore the sync API will redirect to the FITS file result of the underlying async job.
+
+The job representation for a successful async request will list at least two results: the FITS file, and the URL for the output Butler collection that contains both that FITS file and the metadata about the cutout request.
```
Butler doesn't have a concept of a URL to a collection. It only has URLs to datasets.

In OCPS I think they work around this problem by calling `butler retrieve-artifacts` (or equivalent python API) to retrieve all the relevant output files and place them in a temporary location that is available as a URL.

rra replied:
The implication of this is that the image cutout service cannot return a Butler collection, since the only thing that a UWS job can return is a set of URLs. This means that the user will not be able to access or move the Butler collection as a collection, only retrieve the individual files contained in it. I'll add that as an open question if that's okay, or if we'll need to add the concept of a URL to a Butler collection.

timj replied:
We have a convention that the URL includes the run collection name in the root as a "directory" but that's just something that comes out of the file template we define and not something butler or datastore is requiring will happen. In theory the equivalent of `os.path.commonprefix` would return a bucket location that is what you need but as I say, that's pure chance at the moment (but might be true for ever).

rra replied:
Note that this is primarily interesting in a client/server Butler future, since that's presumably what we will go into production with. So if there will be a concept of a URL to a collection with client/server Butler, that's all that's needed, and we can work around it until then. (So perhaps this is a design consideration for client/server Butler.)

timj replied:
It's fundamentally the case that butler datastore doesn't know about collections at all. The registry knows about collections but datastore is just writing files using a template constructed from the DatasetRef that it's been told to write.  Client/server vs direct connection doesn't affect that. As I say above, we could enforce it solely by policy in requiring that every file template starts with a "run" and then from the set of matching URIs we could construct a common prefix that would effectively be the run collection URL. That's something a client could construct if butler itself was enforcing a template-starts-with-run policy.

rra replied:
I think the key point that I'm missing is that I don't understand the API to the Butler and probably need to go do some reading. The mental concept I had was that a collection would be something a client could enumerate the members of and retrieve individual objects from, but it sounds like that mental model may be incorrect.

timj replied:
A client can ask registry for all the datasets in a collection. The client can also ask datastore for the URIs for all the artifacts in those datasets. Armed with that knowledge a client could (if policy allowed) work out a common prefix for all those URIs that would in theory be a URI to the collection.  Butler though does not think of a URI to a collection as a concept itself.

rra replied:
If the request to the registry for all the datasets in a collection is representable as a URL (which presumably it might be with client/server butler?), then maybe we could treat that as the URL to the collection? It at least seems like something reasonable to return as one of the results of an image cutout.

rra replied:
BTW, I should mention that one of the points of making the Butler collection available as a collection rather than as a set of output files was so that someone could more easily transfer the entire Butler collection into their own space and out of the scratch space used for image cutout responses. So I think one of the goals is to provide whatever reference would provide the necessary inputs to `Butler.transfer_from`.

timj replied:
Querying registry for all datasets in the collection is going to (1) need auth, and (2) retrieve a collection of DatsetRef in JSON form (with the URIs inside). We could add a new API for "return URIs (as signed URLs) for this query" but you still need to sort out the auth issue -- is the collection owned by the SODA service? Assuming the SODA service is only runnable by data rights users presumably you could set group ownership for the collection to that user or run the cutout service as that user directly? Is there any way for a user to say that they'd actually want the results to turn up directly in their personal user collection behind the scenes? Would you want the SODA output butler repo to be independent of the main repo? ie do we set up a chained datastore with main repo as input and a completely separate datastore as output and transfer relevant registry entries over -- this would let the services butler be completely independent and might have some administration pluses.

`Butler.transfer_from` takes `DatasetRef`, not URLs, because we need access to the registry information from the source butler.

Bit of a stream of consciousness this morning...

rra replied:
> Assuming the SODA service is only runnable by data rights users presumably you could set group ownership for the collection to that user or run the cutout service as that user directly?

The SODA service will make client/server Butler calls as the user requesting the cutout, using a delegated access token. I'll make that explicit in this document.

> Is there any way for a user to say that they'd actually want the results to turn up directly in their personal user collection behind the scenes?

The SODA specification says this is possible but doesn't define any way to do it. It would therefore have to be a custom parameter, which we would either have to require (which may cause problems with generic SODA clients) or would still have to decide what happens if it's not provided.

> Would you want the SODA output butler repo to be independent of the main repo?

I have no opinion on this. I'm not sure what the trade-offs would be.

> (2) retrieve a collection of DatsetRef in JSON form (with the URIs inside)

I think this is okay. We can define the result of the Butler collection to be this JSON document and provide a URL that would retrieve it. If this is the input to `Butler.transfer_from`, that feels like the right approach. I don't see any obstacle to returning URLs that require authentication.

rra committed:
Add issue about Butler collection URL

rra committed:
Define the content type of FITS results

rra committed:
Define authentication for cutout operations

rra committed:
Redo the provenance data specification

rra committed:
Be clear that signed result URLs may change

rra committed:
Update handling of task storage

We won't need task storage because we have the external database.

rra committed:
Add open issue for SODA's poor result requirements

rra committed:
Add the backend contract

This isn't fully incorporated into the rest of the text.  I can do
that later as another editing pass, but wanted to capture what we
were asking for in the ticket in the tech note as well.

rra committed:
Change terminology to stencil, update API section

Drop the XML handling section since it turns out we don't have to
parse XML.  Add a section on the permission model.

rra committed:
Significant updates to the backend process

Add a new section on the queuing design to avoid needing to add
too much code to a stack container.  Rework and integrate the
interface contract with the pipeline into the document and be
clearer about what we're doing for the initial implementation.
Mention dramatiq-abort as the mechanism for implementing job
aborting.

rra committed:
Work around Sphinx formatting issues

rra committed:
Fix a few more instances of data ID

rra committed:
Be clear there's only one stencil for now

rra committed:
Change the names of the Dramatiq actors

rra committed:
Add a note about DALI-examples

rra committed:
Note we don't support parameter changing

rra committed:
Update for current implementation

Document that we will be using UUIDs to identify source images and
will use something other than ID for identifying virtual data
products.  Lots of other changes to clean up language or make the
description consistent with the implementation.

rra committed:
Relax dependencies

Use the current version of documenteer, since the old version
seems to be causing some weird error in GitHub Actions.
